### PR TITLE
store/tikv/retry: define Config instead of BackoffType

### PR DESCRIPTION
**Pull request by AndreMouche**
Signed-off-by: AndreMouche

### What problem does this PR solve?
This PR defines Backoff Config instead of BackoffType, with this PR, the caller can define any Configuration for backoff. While with BackoffType enum which supports a new type is not very friendly, and makes the retry package not like a package ^-^

Part of https://github.com/pingcap/tidb/issues/22513

### Check List 
Tests 
- Unit test

### Release note 
- No release note

### CODE REVIEW

File: store/tikv/prewrite.go [hunk 1]
```diff
@@ -269,7 +269,7 @@ func (action actionPrewrite) handleSingleBatch(c *twoPhaseCommitter, bo *Backoff
 		}
 		atomic.AddInt64(&c.getDetail().ResolveLockTime, int64(time.Since(start)))
 		if msBeforeExpired > 0 {
-			err = bo.BackoffWithMaxSleep(retry.BoTxnLock, int(msBeforeExpired), errors.Errorf("2PC prewrite lockedKeys: %d", len(locks)))
+			err = bo.BackoffWithCfgAndMaxSleep(retry.BoTxnLock, int(msBeforeExpired), errors.Errorf("2PC prewrite lockedKeys: %d", len(locks)))
```
Comment (disksing) [inline store/tikv/prewrite.go:272]: I think `BackoffWithMaxSleep` is ok. Now all `Backoff*`s need config. We do no need to mention it specially.

File: store/tikv/retry/backoff.go [hunk 2]
```diff
@@ -309,30 +135,30 @@ func (b *Backoffer) BackoffWithMaxSleep(typ BackoffType, maxSleepMs int, err err
 		}
 		logutil.BgLogger().Warn(errMsg)
 		// Use the first backoff type to generate a MySQL error.
-		return b.types[0].(BackoffType).TError()
+		return b.types[0].(*Config).err
```
Comment (disksing) [inline store/tikv/retry/backoff.go:138]: can we changes `types` to `configs []*Config`? It seems `*Config` is the only type.

File: store/tikv/prewrite.go (hunk 1 above)
Comment (AndreMouche) [reply store/tikv/prewrite.go:272]: Sorry, we need to keep the origin BackoffWithMaxSleep since br is using it.

**Review by disksing — APPROVED** [review]

**Review by sticnarf — APPROVED** [review]

**Pull request merged**